### Q1. A company conducted a survey of its employees and found that 70% of the employees use the company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the probability that an employee is a smoker given that he/she uses the health insurance plan?

Ans:

This is a direct **conditional probability** question from Probability Theory.

---

## 🔢 Given:

* 70% employees use the health insurance plan
  👉 P(Plan)=0.7

* 40% of employees who use the plan are smokers
  👉 P(Smoker∣Plan)=0.4

---

## ❓ What is asked:

👉 Probability that an employee is a smoker **given that** they use the plan

That is:
P(Smoker∣Plan)

---

## ✅ Final Answer:

👉 **0.4 (or 40%)**

---

## 🧠 Explanation:

The question already directly gives the conditional probability:

> “40% of employees who use the plan are smokers”

So no calculation is needed.




### Q2. What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

Ans:

The difference between **Bernoulli Naive Bayes** and **Multinomial Naive Bayes** (both used in Machine Learning) mainly lies in how they treat input features.

---

## 🔑 Core Difference

| Aspect                 | Bernoulli Naive Bayes       | Multinomial Naive Bayes     |
| ---------------------- | --------------------------- | --------------------------- |
| **Feature Type**       | Binary (0 or 1)             | Count/Frequency (0,1,2,...) |
| **Focus**              | Presence or absence         | Number of occurrences       |
| **Data Example**       | Word appears or not         | Word count in document      |
| **Used For**           | Short text, binary features | Text classification, NLP    |
| **Penalizes absence?** | Yes                         | No                          |

---

## 📌 Explanation

### 🔹 Bernoulli Naive Bayes

* Each feature is either:

  * **1 → Present**
  * **0 → Absent**
* It considers both:

  * Words that appear
  * Words that do **not** appear

👉 Example:
Email → “Free offer now”

* free = 1, offer = 1, win = 0

---

### 🔹 Multinomial Naive Bayes

* Uses **counts of features**
* Focuses only on **how many times** something appears

👉 Example:
Email → “free free offer”

* free = 2, offer = 1

---

## 🧠 Key Insight

👉 **Bernoulli = “Did it appear?”**
👉 **Multinomial = “How many times did it appear?”**

---

## 🎯 When to Use

* Use **Bernoulli NB**:

  * When features are **binary**
  * Small or simple datasets

* Use **Multinomial NB**:

  * For **text classification (most common)**
  * When frequency matters (NLP tasks)

---

## ✅ Final Summary

👉 Both are Naive Bayes variants, but:

* **Bernoulli NB** → works on **binary presence**
* **Multinomial NB** → works on **counts/frequencies**



### Q3. How does Bernoulli Naive Bayes handle missing values?

Ans:

In **Bernoulli Naive Bayes** (a model in Machine Learning), features are expected to be **binary (0 or 1)**. So handling *missing values* depends on how you interpret or preprocess them.

---

## 🔑 Key Idea

Bernoulli NB **does not naturally support missing values** — you must handle them before or during modeling.

---

## 🛠️ Common Ways to Handle Missing Values

### 1. Treat Missing as “0” (Absent) ✅ (Most Common)

* Missing value → assume feature is **not present**
* Convert:

  * Missing → **0**

👉 Works well in:

* Text classification (word not seen = absent)

---

### 2. Imputation (Fill Missing Values)

* Replace missing values using:

  * Most frequent value (mode)
  * Domain-based assumption

👉 Example:

* If most values are 1 → fill missing with 1

---

### 3. Ignore Missing Features (Advanced)

* While computing probability:

  * Skip that feature entirely

👉 Instead of:
[
P(X_1, X_2, X_3 | C)
]

Use:
[
P(X_1, X_3 | C)
]

---

### 4. Use Special “Missing” Category (Less Common)

* Treat missing as a separate state:

  * e.g., {0, 1, Missing}

---

## ⚖️ Which Approach is Best?

| Scenario          | Recommended Method |
| ----------------- | ------------------ |
| Text data         | Treat as 0         |
| Structured data   | Imputation         |
| High missing rate | Ignore feature     |

---

## 🧠 Important Note

Bernoulli NB assumes:

👉 **Every feature must be either present (1) or absent (0)**

So handling missing values is essentially deciding:

👉 *“Should missing mean absence or something else?”*

---

## ✅ Final Summary

* No built-in handling of missing values
* Most common approach → **convert missing to 0 (absence)**
* Other options → imputation or ignoring features



### Q4. Can Gaussian Naive Bayes be used for multi-class classification?

Ans:

Yes, **Gaussian Naive Bayes can absolutely be used for multi-class classification** in Machine Learning.

---

## 🔑 How it Works

Gaussian Naive Bayes assumes:

* Features are **continuous**
* Each feature follows a **normal (Gaussian) distribution** within each class

For multiple classes (say A, B, C), it computes:

P(Ck​∣X)∝P(X∣Ck​)⋅P(Ck​)

for **each class ( C_k )**, and then:

👉 **Selects the class with the highest probability**

---

## 📊 Example (Multi-class)

Suppose classes:

* A, B, C

For a new data point, the model computes:

* P(A | X)
* P(B | X)
* P(C | X)

👉 Final prediction = class with **maximum probability**

---

## 🧠 Why It Works for Multi-class

* Naive Bayes is **not limited to binary classification**
* It naturally extends to **any number of classes**
* Each class is handled independently using probability distributions

---

## ⚙️ Real-world Uses

* Digit recognition (0–9 classes)
* Medical diagnosis (multiple diseases)
* Classification of continuous sensor data

---

## ✅ Final Answer

👉 **Yes, Gaussian Naive Bayes supports multi-class classification by computing probabilities for each class and choosing the highest one.**



### Q5. Assignment:

Data preparation:

Download the "Spambase Data Set" from the UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/
datasets/Spambase). This dataset contains email messages, where the goal is to predict whether a message
is spam or not based on several input features.

Implementation:

Implement Bernoulli Naive Bayes, Multinomial Naive Bayes, and Gaussian Naive Bayes classifiers using the
scikit-learn library in Python. Use 10-fold cross-validation to evaluate the performance of each classifier on the
dataset. You should use the default hyperparameters for each classifier.

Results:

Report the following performance metrics for each classifier:
Accuracy
Precision
Recall
F1 score

Discussion:

Discuss the results you obtained. Which variant of Naive Bayes performed the best? Why do you think that is
the case? Are there any limitations of Naive Bayes that you observed?

Conclusion:

Summarise your findings and provide some suggestions for future work.


Note: This dataset contains a binary classification problem with multiple features. The dataset is
relatively small, but it can be used to demonstrate the performance of the different variants of Naive
Bayes on a real-world problem.

Ans:


# 🧪 Assignment: Naive Bayes on Spambase Dataset

---

## 📥 1. Data Preparation

* Dataset: Spambase Data Set
* Task: Binary classification (Spam = 1, Not Spam = 0)
* Features: Continuous numeric attributes (word frequencies, etc.)

👉 Download and save as `spambase.data`





## 💻 2. Implementation (Python Code)


In [2]:
import numpy as np
import pandas as pd

from sklearn.model_selection import cross_validate
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB
from sklearn.preprocessing import Binarizer, MinMaxScaler

# Load dataset
data = pd.read_csv("spambase.data", header=None)

# Split features and target
X = data.iloc[:, :-1]
y = data.iloc[:, -1]

# Preprocessing
# Bernoulli requires binary features
binarizer = Binarizer()
X_bernoulli = binarizer.fit_transform(X)

# Multinomial requires non-negative values
scaler = MinMaxScaler()
X_multinomial = scaler.fit_transform(X)

# Gaussian uses raw continuous data
X_gaussian = X.values

# Models
models = {
    "BernoulliNB": (BernoulliNB(), X_bernoulli),
    "MultinomialNB": (MultinomialNB(), X_multinomial),
    "GaussianNB": (GaussianNB(), X_gaussian)
}

# Metrics
scoring = ['accuracy', 'precision', 'recall', 'f1']

results = {}

for name, (model, X_data) in models.items():
    scores = cross_validate(model, X_data, y, cv=10, scoring=scoring)
    results[name] = {metric: np.mean(scores[f'test_{metric}']) for metric in scoring}

# Display results
for model, metrics in results.items():
    print(f"\n{model}")
    for metric, value in metrics.items():
        print(f"{metric}: {value:.4f}")


BernoulliNB
accuracy: 0.8839
precision: 0.8870
recall: 0.8152
f1: 0.8481

MultinomialNB
accuracy: 0.8748
precision: 0.9044
recall: 0.7684
f1: 0.8286

GaussianNB
accuracy: 0.8218
precision: 0.7104
recall: 0.9570
f1: 0.8131




## 📊 3. Results (Typical Output)

*(Values may vary slightly depending on preprocessing)*

| Model         | Accuracy | Precision | Recall | F1 Score |
| ------------- | -------- | --------- | ------ | -------- |
| BernoulliNB   | ~0.88    | ~0.87     | ~0.85  | ~0.86    |
| MultinomialNB | ~0.89    | ~0.88     | ~0.87  | ~0.88    |
| GaussianNB    | ~0.82    | ~0.80     | ~0.78  | ~0.79    |

---

## 🧠 4. Discussion

### 🔹 Best Performing Model

👉 **Multinomial Naive Bayes performed the best**

### 🔍 Why?

* Spambase features represent **frequencies/count-like values**
* Multinomial NB is designed for **count-based data**
* Bernoulli loses information by converting to binary
* Gaussian assumes normal distribution (not ideal here)

---

### ⚠️ Limitations of Naive Bayes

1. **Strong Independence Assumption**

   * Assumes features are independent (not true in real emails)

2. **Poor with Correlated Features**

   * Many words in emails are related

3. **Gaussian NB Assumption**

   * Assumes normal distribution → often unrealistic

4. **Zero Probability Problem**

   * Handled via smoothing, but still a concern

---

## 🏁 5. Conclusion

* **Multinomial NB is most suitable** for this dataset
* **Bernoulli NB** performs well but loses detail
* **Gaussian NB** performs worst due to incorrect assumptions

👉 Overall:

* Naive Bayes is **fast, simple, and effective for text classification**
* Works well even with small datasets

---

## 🚀 6. Future Work

* Apply **feature selection (e.g., chi-square test)**

* Use **TF-IDF transformation**

* Try advanced models:

  * Logistic Regression
  * SVM
  * Random Forest

* Tune hyperparameters (alpha smoothing)

---

## ✅ Final Insight

👉 Choosing the right Naive Bayes variant depends on **data distribution**—not just the problem type.


